### DISCUSSION TITLE
The group velocity of laser

### DISCUSSION THREAD

**Discussion by abecedarian1857** [epochpic/epoch#617] [Q&A]
Hi there,
I want to limit the group velocity of the laser by setting the density of the plasma, which is usually determined by the dispersion relation
 $v~c*(1-w_pe^2/w_0^2)$, where $w_pe$ is the electron plasam frequency, $w_0$ is the laser frequency. For example, in input.desk I use the plasma density equal 0.001 critical density of laser. In theory, the group velocity should be $0.9995*c$, but according to the analysis of the results, it is only $0.9956*c$(This result is calculated by comparing the position of the laser wave packet at different times)。
 For reference here is my input.desk :

#linear

begin:constant

  lambda0 = 1.e-6   #GS
  omega = 2*pi*c/lambda0
  T0 = lambda0/c  #T0=10/3 fs
    
  a_peak = 0.3  #GS 0.04*0.6065=0.0243
  E_peak = a_peak*me*omega*c/ev   #452281426428.7536
  
  n_c = 1.1e15/lambda0/lambda0
  EDEN =0.001*n_c
   
  TUP = 0  #LG
  TPULSE = 1000*T0   #GS 3.3PS,30*T0=100fs#500 * femto
  #TFULL= 4000*T0  #TPULSE+LX/v  #out window
  TFULL= TPULSE
 # fai1= if((y gt 0.0),pi*0.5+atan(z/y),pi*1.5+atan(z/y))
  tshape = gauss(time,3.*6.37*T0,sqrt(2)*6.37*T0)
  profile_ybunch=if((y gt 0e-6)and(y lt 5.0e-6),1,0.0)
############### vac ####
  x1 = 0.e-6
  x2 = 15000.e-6
  y1 = -20.e-6
  y2 = 20.e-6
############### vac ####  
  spot = 9.e-6 #LG
  DN=1.13e26/81 #gradient
 # r = sqrt(z^2+y^2) #3d
  r = sqrt(y^2) #2d  
end:constant

begin:control
  nx = 2600 #lambda0/20
  ny = 280   #lambda0/20
  npart =20*nx*ny
  t_end =  TFULL  #4000*T0 simutime~3.3ps #1000 * femto
  x_min = 0.0
  x_max = 130.0e-6
  y_min = -20.0e-6
  y_max = 20.0e-6
  smooth_currents=T
  smooth_iterations=30
  stdout_frequency=10
end:control

begin:boundaries
  bc_x_min = simple_laser
  bc_x_max = simple_outflow
  bc_y_min = open
  bc_y_max = open
end:boundaries

begin:laser   #GS
  boundary = x_min
  id = 1
  amp = E_peak
  pol_angle = -0.5*pi
  phase = 0.0  #GS:0;  LG:fai1
  omega = omega
  profile = gauss(r,0.0,spot)
#  profile = gauss(r,0.0,spot)*sqrt(2.0)*r/spot #LG
  
  #tshape = if((time gt TUP)and(time lt TFULL), 1, 0.0)
  #tshape = if( (time gt 0.0) and (time lt TFULL) ,sin(time*pi/TFULL)^2,0)
  t_profile = tshape
  t_start = 0.0
  t_end = TPULSE 
end:laser

begin:injector
boundary = x_min
species = electron_1
t_start = 92*T0
t_end  = 93*T0
nparticles_per_cell=9
number_density=1e9*profile_ybunch
temp_ev=0
drift_x=2.675e-19
end:injector

begin:species
   name = electron_1
   charge = -1.0
   mass = 1.0
   bc_x_min=simple_outflow
   bc_x_max=simple_outflow
end:species

begin:species
   name = electron
   charge = -1.0
   mass = 1.0
   frac = 0.5
   density = if((y gt y1)and(y lt y2), EDEN, 0.0)
   density = if((x gt x1)and(x lt x2),EDEN,0.0)
   temp_ev = 0   
end:species

begin:species
   name = ion
   charge = 1.0
   mass = 1836.0
   frac = 0.5
   density = density(electron)
   temp_ev = 0   
end:species

begin:window
move_window =T
window_v_x = 0.9989*c
window_start_time = 100*T0
bc_x_min_after_move = simple_outflow   
bc_x_max_after_move = simple_outflow   
end:window

begin:subset
name = inject
random fraction=1
include_species:electron_1
y_min=-5e-6
y_max=5e-6
persist_start_time=92*T0
end:subset

begin:output
  name = out1
  file_prefix = field
 #time_start=50*T0
  dt_snapshot =50*T0  ##30.0*T0=100fs# 20 * femto~6*T0
  #restart_dump_every = 5
  grid = always + single
  ex = always + single
  ey = always + single
  ez = always + single
 # bx = always + single
#  by = always + single
#  bz = always + single
  #particles = always + single
  #vx = always+single
  #vy = always+single
  #px = always + single
  #py = always + single
  #pz = always + single
  #ekbar=always + species +single
  #number_density = always + species +single +no_sum
  #particle_weight = always +single
  #gamma=always+single
end:output

begin:output
  name = out2
  file_prefix = dens
  # time_start=50*T0
  dt_snapshot =50* T0  ##30.0*T0=100fs# 20 * femto~6*T0
  #restart_dump_every = 5
  grid = always + single
  number_density = always + species +single +no_sum
  #ex = always + single
  #ey = always + single
  #ez = always + single
  #bx = always + single
  #by = always + single
  #bz = always + single
  #particles = always + single
  #vx = always+single
  #vy = always+single
  #px = always + single
  #py = always + single
 end :output
    
 begin:output
   name = out3
   file_prefix = position
   time_start=92*T0
   dt_snapshot = T0
   id= inject
   particles =  inject
   vx =  inject
   vy =  inject
   px= inject
   py= inject
end:output

So how to make the simulation result as close to the theoretical value as possible?
I would really appreciate any help or recommendations for resources on how to do this.

Many thanks

**Comment by DanRRRR** [comment]
abecedarian1857
I am fighting approximately with the same problem right now. At the densities like in your case the laser pulse propagates whopping 1.5% faster than the speed of light. Would like to compare with your case to see may be  i am missing something but as you see GitHub distorts the posts. Can you please post again the init.deck a bit different way? The trick when the text will be not distorted is the following:   at the end of your post on a single line  just add three grave accents (on keyboard it is first key, just before key 1 and ! )  and on the next lines place your text and post it.

**Reply by abecedarian1857** [reply]
DanRRR
Thank you for the reminder. And here is my input.desk

~~~
begin:constant

  lambda0 = 1.e-6  
  omega = 2*pi*c/lambda0
  T0 = lambda0/c 
    
  a_peak = 0.3
  E_peak = a_peak*me*omega*c/ev 
  
  n_c = 1.1e15/lambda0/lambda0
  EDEN =0.001*n_c
   
  TUP = 0 
  TPULSE = 1000*T0  
  TFULL= TPULSE
  tshape = gauss(time,3.*6.37*T0,sqrt(2)*6.37*T0)
  profile_ybunch=if((y gt 0e-6)and(y lt 5.0e-6),1,0.0)
  x1 = 0.e-6
  x2 = 15000.e-6
  y1 = -20.e-6
  y2 = 20.e-6
  spot = 9.e-6 
  DN=1.13e26/81 
  r = sqrt(y^2) 
end:constant


begin:control
  nx = 2600 
  ny = 280   
  npart =20*nx*ny
  t_end =  TFULL
  x_min = 0.0
  x_max = 130.0e-6
  y_min = -20.0e-6
  y_max = 20.0e-6
  smooth_currents=T
  smooth_iterations=30
  stdout_frequency=10
end:control

begin:boundaries
  bc_x_min = simple_laser
  bc_x_max = simple_outflow
  bc_y_min = open
  bc_y_max = open
end:boundaries

begin:laser   
  boundary = x_min
  id = 1
  amp = E_peak
  pol_angle = -0.5*pi
  phase = 0.0  
  omega = omega
  profile = gauss(r,0.0,spot)
  t_profile = tshape
  t_start = 0.0
  t_end = TPULSE 
end:laser

begin:injector
  boundary = x_min
  species = electron_1
  t_start = 92*T0
  t_end  = 93*T0
  nparticles_per_cell=9
  number_density=1e9*profile_ybunch
  temp_ev=0
  drift_x=2.675e-19
end:injector

begin:species
   name = electron_1
   charge = -1.0
   mass = 1.0
   bc_x_min=simple_outflow
   bc_x_max=simple_outflow
end:species


begin:species
   name = electron
   charge = -1.0
   mass = 1.0
   frac = 0.5
   density = if((y gt y1)and(y lt y2),EDEN, 0.0)
   density = if((x gt x1)and(x lt x2),EDEN, 0.0)
   temp_ev = 0   
end:species


begin:species
   name = ion
   charge = 1.0
   mass = 1836.0
   frac = 0.5
   density = density(electron)
   temp_ev = 0   
end:species


begin:window
move_window =T
window_v_x = 0.9989*c
window_start_time = 100*T0
bc_x_min_after_move = simple_outflow   
bc_x_max_after_move = simple_outflow   
end:window

begin:subset
name = inject
random fraction=1
include_species:electron_1
y_min=-5e-6
y_max=5e-6
persist_start_time=92*T0
end:subset

begin:output
  name = out1
  file_prefix = field
  dt_snapshot =50*T0
  grid = always + single
  ex = always + single
  ey = always + single
  ez = always + single
end:output

begin:output
  name = out2
  file_prefix = dens
  dt_snapshot =50* T0  
  grid = always + single
  number_density = always + species +single +no_sum
 end :output
    
 begin:output
   name = out3
   file_prefix = position
   time_start=92*T0
   dt_snapshot = T0
   id= inject
   particles =  inject
   vx =  inject
   vy =  inject
   px= inject
   py= inject
end:output

**Comment by DanRRRR** [comment]
[abecedarian1857](https://github.com/abecedarian1857),
Thanks.
Unfortunately by some reason your initial settings cause my code to stall at the time of injector start  (I little bit edited EPOCH from the default for it to be more convenient but still have not resolved some issues this sometimes causes)

**Reply by abecedarian1857** [reply]
Hi DanRRR,
I am so sorry that i forot  to remind you i have recompiled the source code because i want to realize  tracking specific particles as they move through a simulation. It can  compiled in using the “-DPARTICLE ID” compiler flag. But we can remove the relevent code in input.desk. So here is my new input.desk, and it's  group velocity of laser is $0.9970*c$,  the group velocity should be $0.9995c$.
~~~
begin:constant

  lambda0 = 1.e-6  
  omega = 2*pi*c/lambda0
  T0 = lambda0/c 
    
  a_peak = 0.3
  E_peak = a_peak*me*omega*c/ev 
  
  n_c = 1.1e15/lambda0/lambda0
  EDEN =0.001*n_c
   
  TUP = 0 
  TPULSE = 2000*T0  
  TFULL= TPULSE
  tshape = gauss(time,3.*6.37*T0,sqrt(2)*6.37*T0)
  profile_ybunch=if((y gt 0e-6)and(y lt 5.0e-6),1,0.0)
  x1 = 0.e-6
  x2 = 15000.e-6
  y1 = -20.e-6
  y2 = 20.e-6
  spot = 9.e-6 
  DN=1.13e26/81 
  r = sqrt(y^2) 
end:constant


begin:control
  nx = 2600 
  ny = 280   
  npart =20*nx*ny
  t_end =  TFULL
  x_min = 0.0
  x_max = 130.0e-6
  y_min = -20.0e-6
  y_max = 20.0e-6
  smooth_currents=T
  smooth_iterations=30
  stdout_frequency=10
end:control

begin:boundaries
  bc_x_min = simple_laser
  bc_x_max = simple_outflow
  bc_y_min = open
  bc_y_max = open
end:boundaries

begin:laser   
  boundary = x_min
  id = 1
  amp = E_peak
  pol_angle = -0.5*pi
  phase = 0.0  
  omega = omega
  profile = gauss(r,0.0,spot)
  t_profile = tshape
  t_start = 0.0
  t_end = TPULSE 
end:laser



begin:species
   name = electron
   charge = -1.0
   mass = 1.0
   frac = 0.5
   density = if((y gt y1)and(y lt y2),EDEN, 0.0)
   density = if((x gt x1)and(x lt x2),EDEN, 0.0)
   temp_ev = 0   
end:species


begin:species
   name = ion
   charge = 1.0
   mass = 1836.0
   frac = 0.5
   density = density(electron)
   temp_ev = 0   
end:species


begin:window
move_window =T
window_v_x = 0.9989*c
window_start_time = 100*T0
bc_x_min_after_move = simple_outflow   
bc_x_max_after_move = simple_outflow   
end:window



begin:output
  name = out1
  file_prefix = field
  dt_snapshot =50*T0
  grid = always + single
  ex = always + single
  ey = always + single
  ez = always + single
end:output

begin:output
  name = out2
  file_prefix = dens
  dt_snapshot =50* T0  
  grid = always + single
  number_density = always + species +single +no_sum
 end :output

**Comment by DanRRRR** [comment]
Looks strange so far, but preliminary (still waiting in the queue for the longer run) your setup does not go beyond speed of light...
